pkg/cloudprovider/vsphere/nodemanager_test.go

```diff
@@ -0,0 +1,174 @@
+limitations under the License.
+*/
+
+package vsphere
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/simulator"
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestRegUnregNode(t *testing.T) {
+	cfg, ok := configFromEnvOrSim()
+	defer ok()
+
+	vsphere, err := newVSphere(cfg)
+	if err != nil {
+		t.Fatalf("Failed to construct/authenticate vSphere: %s", err)
+	}
+
+	nm := NodeManager{
+		vsphereInstanceMap: vsphere.vsphereInstanceMap,
+		nodeNameMap:        make(map[string]*NodeInfo),
+		nodeUUIDMap:        make(map[string]*NodeInfo),
+		nodeRegUUIDMap:     make(map[string]*v1.Node),
+	}
+
+	vm := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine)
+	name := vm.Config.GuestFullName
+	UUID := vm.Config.Uuid
+	k8sUUID := nm.convertK8sUUIDtoNormal(UUID)
+
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Status: v1.NodeStatus{
+			NodeInfo: v1.NodeSystemInfo{
+				SystemUUID: k8sUUID,
+			},
+		},
+	}
+
+	nm.RegisterNode(node)
+
+	if len(nm.nodeNameMap) != 1 {
+		t.Errorf("Failed: nodeNameMap should be a length of 1")
+	}
+	if len(nm.nodeUUIDMap) != 1 {
+		t.Errorf("Failed: nodeUUIDMap should be a length of  1")
+	}
+	if len(nm.nodeRegUUIDMap) != 1 {
+		t.Errorf("Failed: nodeRegUUIDMap should be a length of  1")
+	}
+
+	nm.UnregisterNode(node)
+
+	if len(nm.nodeNameMap) != 1 {
+		t.Errorf("Failed: nodeNameMap should be a length of  1")
+	}
+	if len(nm.nodeUUIDMap) != 1 {
+		t.Errorf("Failed: nodeUUIDMap should be a length of  1")
+	}
+	if len(nm.nodeRegUUIDMap) != 0 {
+		t.Errorf("Failed: nodeRegUUIDMap should be a length of 0")
+	}
+}
+
+type SearchIndex struct {
+	*simulator.SearchIndex
+	vm *simulator.VirtualMachine
+}
+
+func (s *SearchIndex) FindByDnsName(req *types.FindByDnsName) soap.HasFault {
+	res := &methods.FindByDnsNameBody{Res: new(types.FindByDnsNameResponse)}
+	if req.VmSearch && strings.ToLower(req.DnsName) == strings.ToLower(s.vm.Config.GuestFullName) {
+		res.Res.Returnval = &s.vm.Self
+	}
+	return res
+}
+
+func TestDiscoverNodeByName(t *testing.T) {
+	cfg, ok := configFromEnvOrSim()
+	defer ok()
+
+	vsphere, err := newVSphere(cfg)
+	if err != nil {
+		t.Errorf("Failed to construct/authenticate vSphere: %s", err)
+	}
+
+	nm := NodeManager{
+		vsphereInstanceMap: vsphere.vsphereInstanceMap,
+		nodeNameMap:        make(map[string]*NodeInfo),
+		nodeUUIDMap:        make(map[string]*NodeInfo),
+		nodeRegUUIDMap:     make(map[string]*v1.Node),
+	}
+
+	vm := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine)
+	name := vm.Config.GuestFullName
+
+	var cancel context.CancelFunc
+	for _, instance := range vsphere.vsphereInstanceMap {
+		var ctx context.Context
+		ctx, cancel = context.WithCancel(context.Background())
```

Comment:
No need for `WithCancel` here.  Also note in general you'd want to `defer cancel()` here, rather than outside the loop.

Reply:
You can also use the instance directly as the other tests do:
```go
	instance, ok := vsphere.vsphereInstanceMap[cfg.Global.VCenterIP]
	if !ok {
		t.Fatalf("Couldn't get vSphere instance: %s", cfg.Global.VCenterIP)
	}
```